presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +582,186 @@ private static NestedField nestedField(String name)
+                            for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                                if (!addLambdaSubfields(call, context, lambdaDescriptor, argumentIndexToLambdaSubfieldsMap)) {
+                                    context.setLambdaSubfields(Optional.empty());
+                                    return null;
+                                }
+                            }
+                        }
+                        else if (!functionDescriptor.isAccessingInputValues() && !functionDescriptor.isReturningInputValues()) {
+                            addNoneSubfield(call, argumentIndexToLambdaSubfieldsMap);
+                        }
+                        for (int callArgumentIndex = 0; callArgumentIndex < call.getArguments().size(); callArgumentIndex++) {
+                            if (argumentIndexToLambdaSubfieldsMap.containsKey(callArgumentIndex)) {
+                                context.setLambdaSubfields(Optional.of(argumentIndexToLambdaSubfieldsMap.get(callArgumentIndex)));
+                            }
+                            else {
+                                context.setLambdaSubfields(Optional.empty());
+                            }
+                            call.getArguments().get(callArgumentIndex).accept(this, context);
+                        }
+                        context.setLambdaSubfields(lambdaSubfieldsOriginal);
+                        return null;
+                    }
                 }
-                else {
+                call.getArguments().forEach(argument -> argument.accept(this, context));
+                return null;
+            }
+
+            private static void addNoneSubfield(CallExpression call, Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap)
+            {
+                for (Integer callArgumentIndex : argumentIndexToLambdaSubfieldsMap.keySet()) {
+                    RowExpression argument = call.getArguments().get(callArgumentIndex);
+                    if ((argument.getType() instanceof ArrayType && ((ArrayType) argument.getType()).getElementType() instanceof RowType) ||
+                            (argument.getType() instanceof MapType && ((MapType) argument.getType()).getValueType() instanceof RowType)) {
+                        argumentIndexToLambdaSubfieldsMap.put(callArgumentIndex, ImmutableSet.of(new Subfield("", ImmutableList.of(allSubscripts(), noneSubfield()))));
+                    }
+                }
+            }
+
+            private boolean addLambdaSubfields(CallExpression call, Context context, LambdaDescriptor lambdaDescriptor,
+                    Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap)
+            {
+                if (lambdaDescriptor.getCallArgumentIndex() >= 0 &&
+                        (lambdaDescriptor.getCallArgumentIndex() >= call.getArguments().size() ||
```

Comment:
shouldn't these 2 cases be invalid? Maybe we should add a check to ComplexTypeFunctionDescriptor to ensure this doesn't exist.

Reply:
`ComplexTypeFunctionDescriptor` does not have any information about function signature.